upstream: Fallback policy per selector support

nezdolik opened the pull request:
Signed-off-by: Kateryna Nezdolii <nezdolik@spotify.com>

Duplicate of https://github.com/envoyproxy/envoy/pull/6124 but with completed development. 

Description: This code change allows to redefine fallback policy per specific subset selector. Because of how existing `LbSubsetMap` trie data structure is organised  (mapping subset key to values), is not possible to do lookups for fallback policy only based on subset keys (had to introduce additional trie that maps subset keys to keys and has fallback policy on leaf level). Additional `LbSubsetSelectorFallbackPolicy` enum required to correctly identify the case when fallback policy is not set for given selector (otherwise it would always default to NO_FALLBACK, breaking backwards compatibility, if field is not set we should use top level fallback policy instead). 

Risk Level: Medium
Testing: Done
Docs Changes: Updated related docs
Release Notes:
Fixes #5130

htuch commented:
Please merge master to pick up https://github.com/envoyproxy/envoy/pull/7090

yanavlasov commented on include/envoy/upstream/load_balancer_type.h:
```diff
@@ -17,6 +19,11 @@ namespace Upstream {
  */
 enum class LoadBalancerType { RoundRobin, LeastRequest, Random, RingHash, OriginalDst, Maglev };
 
+typedef std::pair<
```
Any specific reason for using std::pair instead of struct here? Could be more readable to use descriptive names instead of first and second.

nezdolik replied:
@yanavlasov thanks for hint, will switch to struct instead

nezdolik replied:
switched to struct instead

yanavlasov commented:
Please add unit tests for the new functionality in the subset_lb_test.cc

yanavlasov commented on source/common/upstream/subset_lb.cc:
```diff
@@ -110,6 +112,62 @@ void SubsetLoadBalancer::refreshSubsets(uint32_t priority) {
+    unsigned long pos = 0;
+    selectors = selectors_;
+    for (const auto& key : selector_keys) {
+      const auto& selector_it = selectors->subset_keys.find(key);
```
This must be crashing some unit tests as selectors_ is default initialized only.

nezdolik replied:
thanks, fixed, previous version of code looked like this:
<code>
struct SubsetSelectorMap {
    std::unordered_map<std::string, SubsetSelectorMap> subset_keys;
    ...
  }
;</code>
and clang compiler on my machine did not complain about `incomplete type` but build failed on CI. Switched to smart pointers in `SubsetSelectorMap` struct instead.

yanavlasov replied:
Sorry again, I'm unsure how I have missed selectors_ initialization on the first line of this function. Please disregard this comment.

nezdolik replied:
@yanavlasov selectors_ initialization was not there indeed, apologies i had to do force push to correct DCO.

dio commented:
@nezdolik seems like you need to fix DCO. Please follow this: https://github.com/envoyproxy/envoy/blob/master/CONTRIBUTING.md#fixing-dco

To avoid missing DCO in a pushed commit in the future seems like enabling the commit hook (`./support/bootstrap`) is not a bad idea https://github.com/envoyproxy/envoy/blob/master/CONTRIBUTING.md#dco-sign-your-work

yanavlasov commented:
Please add user documentation for the new feature here https://github.com/envoyproxy/envoy/blob/master/docs/root/intro/arch_overview/load_balancing/subsets.rst

yanavlasov commented on include/envoy/upstream/load_balancer_type.h:
```diff
@@ -17,6 +19,11 @@ namespace Upstream {
 
+typedef std::pair<
+    std::set<std::string>,
+    absl::optional<envoy::api::v2::Cluster::LbSubsetConfig::LbSubsetSelectorFallbackPolicy>>
```
Does this actually need to be absl::optional<>? The default value of this enum is NOT_DEFINED_PER_SELECTOR which if I read the code correctly is equivalent to the "empty" optional value.

yanavlasov commented on source/common/upstream/subset_lb.cc:
```diff
@@ -110,6 +112,63 @@ void SubsetLoadBalancer::refreshSubsets(uint32_t priority) {
 }
 
+void SubsetLoadBalancer::initSubsetSelectorMap() {
+  selectors_ = std::make_shared<SubsetSelectorMap>();
```
Since selectors_ is always initialized, I suggest moving its initialization to the constructor initializer list or better yet initialize it right in the header at the point of definition.

nezdolik replied:
@yanavlasov cannot init selectors_ in the header, as it depends on field `subset_selectors_` being initialised in constructor. Moved out this initialisation into separate method, as constructor body is already big and it could be hard to read. What do you think?

yanavlasov replied:
I was suggesting to move just the first line where the pointer is initialized to point to an empty SubsetSelectorMap object to the header. However the way you have is fine as well.

yanavlasov commented on source/common/upstream/subset_lb.h:
```diff
@@ -177,8 +198,14 @@ class SubsetLoadBalancer : public LoadBalancer, Logger::Loggable<Logger::Id::ups
   LbSubsetEntryPtr fallback_subset_;
   LbSubsetEntryPtr panic_mode_subset_;
 
+  absl::optional<LbSubsetEntryPtr> selector_fallback_subset_any_;
```
It is probably overkill to use optional with a pointer value as nullptr pointer value is typically used to indicate that the pointer points to "nothing".

yanavlasov commented on source/common/upstream/subset_lb.h:
```diff
@@ -111,10 +119,19 @@ class SubsetLoadBalancer : public LoadBalancer, Logger::Loggable<Logger::Id::ups
   typedef std::unordered_map<HashedValue, LbSubsetEntryPtr> ValueSubsetMap;
   typedef std::unordered_map<std::string, ValueSubsetMap> LbSubsetMap;
 
+  struct SubsetSelectorMap {
```
I think it is fine as it is. However you could likely simplify this. The subset fallback policy is fully identified by the set of selector keys for this subset. You can concatenate the sorted selector keys into one string and use it as a key for finding the fallback policy. In this way you do not need to build a trie and can just use std::unordered_set<std::string, envoy::api::v2::Cluster::LbSubsetConfig::LbSubsetSelectorFallbackPolicy>

nezdolik replied:
@yanavlasov agree that that way code would be simplified. Was thinking about tradeoff though, if to switch to plain map, on each lookupHosts call LB will have to perform concatenation of keys from match criteria to look it up in map. While first approach brings code complexity but is kind of precomputing things on LB object creation so that when lookupHosts called we just iterate over each match criteria key and look it up recursively in a trie. If it's not a big deal of match criteria keys concatenation (from memory usage perspective), then having plain map seems more beneficial.

yanavlasov commented on source/common/upstream/subset_lb.cc:
```diff
@@ -110,6 +112,63 @@ void SubsetLoadBalancer::refreshSubsets(uint32_t priority) {
+void SubsetLoadBalancer::initSelectorFallbackSubset(
+    const absl::optional<envoy::api::v2::Cluster::LbSubsetConfig::LbSubsetSelectorFallbackPolicy>&
+        fallback_policy) {
+  if (fallback_policy == envoy::api::v2::Cluster::LbSubsetConfig::NO_FALLBACK_PER_SELECTOR) {
```
The first if statement is superfluous.

yanavlasov commented on source/common/upstream/subset_lb.cc:
```diff
@@ -110,6 +112,63 @@ void SubsetLoadBalancer::refreshSubsets(uint32_t priority) {
+        selectors = childSelector->second;
+      } else {
+        selectors = selector_it->second;
+      };
```
superfluous semicolon

yanavlasov commented on source/common/upstream/subset_lb.h:
```diff
@@ -177,8 +198,14 @@ class SubsetLoadBalancer : public LoadBalancer, Logger::Loggable<Logger::Id::ups
   LbSubsetEntryPtr panic_mode_subset_;
 
+  absl::optional<LbSubsetEntryPtr> selector_fallback_subset_any_;
+  absl::optional<LbSubsetEntryPtr> selector_fallback_subset_default_;
```
I think the ownership of the above pointers is never actually shared. It does not hurt for them to be shared_ptr, but it is more readable to use std::unique_ptr to clearly indicate that this member is only owned by the parent SubsetLoadBalancer object.

yanavlasov commented on test/common/upstream/subset_lb_test.cc:
```diff
@@ -1628,6 +1801,201 @@ TEST_P(SubsetLoadBalancerTest, GaugesUpdatedOnDestroy) {
+  EXPECT_CALL(subset_info_, fallbackPolicy())
+      .WillRepeatedly(Return(envoy::api::v2::Cluster::LbSubsetConfig::NO_FALLBACK));
+
+  std::vector<SubsetSelector> subset_selectors = {
```
Please add selectors with multiple keys. Otherwise you do not really exercise building trie deeper than 1.

nezdolik replied:
added test with compound selectors

nezdolik commented:
@yanavlasov addressed some of review comments (added testing for compound selector, switched to struct for `SubsetSelector` type, removed optional in struct), all others are in progress

htuch commented on api/envoy/api/v2/cds.proto:
```diff
@@ -344,6 +344,19 @@ message Cluster {
+    // etc). If DEFAULT_SUBSET is selected, load balancing is performed over the
+    // endpoints matching the values from the default_subset field.
+    enum LbSubsetSelectorFallbackPolicy {
+      NOT_DEFINED_PER_SELECTOR = 0;
```
I think the `_PER_SELECTOR` suffix is a bit verbose/redundant here, or maybe I'm missing something, LMK.

htuch replied:
I see, this is because we already have these values above for the top-level. Can you nest this `enum` in `LbSubsetSelector`? It probably belongs there anyway.

nezdolik replied:
After working on doc part, agree that is less confusing from user perspective to get rid of `_PER_SELECTOR` and is now possible after moving enum into `LbSubsetSelector`.

htuch commented on api/envoy/api/v2/cds.proto:
```diff
@@ -344,6 +344,19 @@ message Cluster {
       DEFAULT_SUBSET = 2;
     }
 
+    // If NOT_DEFINED top level config fallback policy is used instead.
```
Maybe move these comments above each respective enum value definition line.

nezdolik replied:
@htuch thanks, api level comments fixed

htuch commented on api/envoy/api/v2/cds.proto:
```diff
@@ -363,6 +363,24 @@ message Cluster {
+      // Allows to override top level fallback policy per selector.
+      enum LbSubsetSelectorFallbackPolicy {
+        // If NOT_DEFINED top level config fallback policy is used instead.
+        NOT_DEFINED_PER_SELECTOR = 0;
```
Were you going to drop `_PER_SELECTOR`?

nezdolik replied:
@htuch just pushed the commit + updated docs

nezdolik commented on source/common/upstream/load_balancer_impl.h:
```diff
@@ -13,6 +13,8 @@
 #include "common/protobuf/utility.h"
 #include "common/upstream/edf_scheduler.h"
 
+#include "absl/types/optional.h"
```
looks like this is not needed anymore

nezdolik commented:
@yanavlasov addressed all the comments (except selectors_ data structure type, that raised a discussion), please take a look

nezdolik commented:
/retest

yanavlasov commented:
Looks good to me. Excellent work.

htuch commented on api/envoy/api/v2/cds.proto:
```diff
@@ -367,6 +367,24 @@ message Cluster {
     message LbSubsetSelector {
       // List of keys to match with the weighted cluster metadata.
       repeated string keys = 1;
+      // The behavior used when no endpoint subset matches the selected route's
```
Can you add release notes in version_history.rst?

nezdolik replied:
yes

htuch commented on docs/root/intro/arch_overview/load_balancing/subsets.rst:
```diff
@@ -18,7 +18,8 @@ specifies no metadata or no subset matching the metadata exists, the subset load
 the cluster had no hosts. Conversely, the ``ANY_ENDPOINT`` fallback policy load balances across all
 hosts in the cluster, without regard to host metadata. Finally, the ``DEFAULT_SUBSET`` causes
-fallback to load balance among hosts that match a specific set of metadata.
+fallback to load balance among hosts that match a specific set of metadata. Is possible to
```
Nit: s/Is possible/It is possible/

htuch commented on include/envoy/upstream/load_balancer_type.h:
```diff
@@ -25,6 +25,14 @@ enum class LoadBalancerType {
 };
 
+struct SubsetSelector {
+  std::set<std::string> selector_keys;
```
Is there a reason this set is ordered?

nezdolik replied:
@htuch yes, that's how we ensure fast lookups in subset selectors for given metadata match criteria (which is sorted as well), see also https://github.com/envoyproxy/envoy/blob/master/include/envoy/router/router.h#L514

htuch commented on include/envoy/upstream/load_balancer_type.h:
```diff
@@ -25,6 +25,14 @@ enum class LoadBalancerType {
+struct SubsetSelector {
+  std::set<std::string> selector_keys;
+  envoy::api::v2::Cluster::LbSubsetConfig::LbSubsetSelector::LbSubsetSelectorFallbackPolicy
+      fallback_policy;
```
Nit: both these fields should be `_` suffixed, e.g. `fallback_policy_`.

htuch commented on include/envoy/upstream/load_balancer_type.h:
```diff
@@ -25,6 +25,14 @@ enum class LoadBalancerType {
+      fallback_policy;
+};
+
+typedef std::shared_ptr<SubsetSelector> SubsetSelectorPtr;
```
Nit: most code is now preferring `using` to `typedef` for this.

nezdolik replied:
that's a new keyword i was not aware of :)

htuch commented on source/common/upstream/subset_lb.cc:
```diff
@@ -110,6 +112,63 @@ void SubsetLoadBalancer::refreshSubsets(uint32_t priority) {
+  SubsetSelectorMapPtr selectors;
+  for (const auto& subset_selector : subset_selectors_) {
+    const auto& selector_keys = subset_selector->selector_keys;
+    unsigned long pos = 0;
```
Nit: Envoy style prefers explicit integer size types, e.g. `uint32_t`.

htuch commented on source/common/upstream/subset_lb.cc:
```diff
@@ -110,6 +112,63 @@ void SubsetLoadBalancer::refreshSubsets(uint32_t priority) {
+      pos++;
+      if (selector_it == selectors->subset_keys.end()) {
+        selectors->subset_keys.emplace(std::make_pair(key, std::make_shared<SubsetSelectorMap>()));
+        const auto& childSelector = selectors->subset_keys.find(key);
```
Nit: camel case, `child_selector`

nezdolik commented:
/retest

rgs1 commented:
@nezdolik should panic mode also be configurable per selector?

zuercher commented on source/common/upstream/subset_lb.cc:
```diff
@@ -141,6 +211,65 @@ HostConstSharedPtr SubsetLoadBalancer::chooseHost(LoadBalancerContext* context)
+
+HostConstSharedPtr SubsetLoadBalancer::chooseHostForSelectorFallbackPolicy(
+    const absl::optional<
+        envoy::api::v2::Cluster::LbSubsetConfig::LbSubsetSelector::LbSubsetSelectorFallbackPolicy>&
```
I don't think you need absl::optional here. It's only called if fallback policy is set.

zuercher commented on source/common/upstream/subset_lb.cc:
```diff
@@ -118,6 +177,17 @@ HostConstSharedPtr SubsetLoadBalancer::chooseHost(LoadBalancerContext* context)
+      HostConstSharedPtr host =
+          chooseHostForSelectorFallbackPolicy(selector_fallback_policy, context);
+      // Subset lookup succeeded, return this result even if it's nullptr.
+      return host;
```
nit: elide the variable and just `return chooseHostForSelectorFallbackPolicy(...)`

zuercher commented on source/common/upstream/subset_lb.cc:
```diff
@@ -141,6 +211,65 @@ HostConstSharedPtr SubsetLoadBalancer::chooseHost(LoadBalancerContext* context)
+        selector_fallback_subset_any_->priority_subset_->lb_->chooseHost(context);
+    if (host != nullptr) {
+      return host;
+    }
```
nit: elide local variable and return return the result of chooseHost (same below).

zuercher commented on source/common/upstream/subset_lb.h:
```diff
@@ -111,10 +119,20 @@ class SubsetLoadBalancer : public LoadBalancer, Logger::Loggable<Logger::Id::ups
   typedef std::unordered_map<std::string, ValueSubsetMap> LbSubsetMap;
 
+  struct SubsetSelectorMap {
+    std::unordered_map<std::string, SubsetSelectorMapPtr> subset_keys;
```
nit: field names should end in an underscore (_) (fallback_policy, too)

zuercher commented on include/envoy/upstream/load_balancer_type.h:
```diff
@@ -25,6 +25,14 @@ enum class LoadBalancerType {
+      fallback_policy_;
+};
+
+using SubsetSelectorPtr = std::shared_ptr<SubsetSelector>;
```
nit: I think our style is to use a typedef here.

zuercher replied:
Oops. I didn't @htuch's earlier comment re: using. I shouldn't have asked you to change it. Feel free to leave it as typedef.

nezdolik replied:
@zuercher switched back to `using` :)

zuercher replied:
Thank you!

zuercher commented on test/common/upstream/subset_lb_test.cc:
```diff
@@ -1628,6 +1704,210 @@ TEST_P(SubsetLoadBalancerTest, GaugesUpdatedOnDestroy) {
+
+  std::vector<SubsetSelectorPtr> subset_selectors = {
+      std::make_shared<SubsetSelector>(SubsetSelector{
+          {"version"}, envoy::api::v2::Cluster::LbSubsetConfig::LbSubsetSelector::ANY_ENDPOINT})};
```
Maybe make this NO_FALLBACK to distinguish between the case where we use this fallback (context_unknown_key) vs. when the top level one is used (context_unknown_value)?

nezdolik replied:
fixed

zuercher commented on api/envoy/api/v2/cds.proto:
```diff
@@ -367,6 +367,24 @@ message Cluster {
+          [(validate.rules).enum.defined_only = true];
+
+      // Allows to override top level fallback policy per selector.
+      enum LbSubsetSelectorFallbackPolicy {
```
I wish there was a way to reuse the other enum. Oh well.

zuercher commented on source/common/upstream/subset_lb.cc:
```diff
@@ -110,6 +112,63 @@ void SubsetLoadBalancer::refreshSubsets(uint32_t priority) {
+    const auto& selector_keys = subset_selector->selector_keys_;
+    uint32_t pos = 0;
+    selectors = selectors_;
+    for (const auto& key : selector_keys) {
```
What do you think about skipping this loop if `subset_selector->fallback_policy_` is undefined? As it stands, the default behavior will be to build up this data structure with all the subsets even if no subset has a custom fallback policy configured. If you skip adding the subsets with no custom fallback, then `SubsetSelectorMap.fallback_policy_` doesn't need to be optional and you can get rid of that complicated ternary below.

nezdolik replied:
agree, thanks for pointing out

nezdolik commented:
@rgs1 possibly yes, with this change in case there is override per selector and no hosts are found, `nullptr` is returned.

rgs1 commented on source/common/upstream/subset_lb.cc:
```diff
@@ -110,6 +112,60 @@ void SubsetLoadBalancer::refreshSubsets(uint32_t priority) {
+    HostPredicate predicate = [](const Host&) -> bool { return true; };
+    selector_fallback_subset_any_ = std::make_unique<LbSubsetEntry>();
+    selector_fallback_subset_any_->priority_subset_.reset(
+        new PrioritySubsetImpl(*this, predicate, locality_weight_aware_, scale_locality_weight_));
```
nit: use `make_shared()` for consistency?

rgs1 commented on source/common/upstream/subset_lb.cc:
```diff
@@ -110,6 +112,60 @@ void SubsetLoadBalancer::refreshSubsets(uint32_t priority) {
+                                        default_subset_metadata_, std::placeholders::_1);
+    selector_fallback_subset_default_ = std::make_unique<LbSubsetEntry>();
+    selector_fallback_subset_default_->priority_subset_.reset(
+        new PrioritySubsetImpl(*this, predicate, locality_weight_aware_, scale_locality_weight_));
```
ditto

nezdolik commented:
@rgs1 switched back to `shared_ptr`, ideally `LbSubsetEntryPtr` should be `unique_ptr`, not sure why it was declared shared in first place

zuercher commented:
@nezdolik if you can get that merge conflict fixed we can merge this (and then @snowp can merge his subset lb change).

nezdolik commented:
@zuercher merge conflict should be fixed now
